Mathlib/Topology/UniformSpace/Ultra/Prod.lean

```diff
@@ -0,0 +1,74 @@
+/-
+Copyright (c) 2025 Yakov Pechersky. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yakov Pechersky
+-/
+import Mathlib.Topology.UniformSpace.Basic
+import Mathlib.Topology.UniformSpace.Ultra.Basic
+
+/-!
+# Binary products of ultrametric (nonarchimedean) uniform spaces
+
+## Main results
+
+* `IsUltraUniformity.prod`: a product of uniform spaces with nonarchimedean uniformities
+  has a nonarchimedean uniformity.
+
+-/
+
+variable {X Y : Type*}
+
+lemma IsTransitiveRel.entourageProd {s : Set (X × X)} {t : Set (Y × Y)}
+    (hs : IsTransitiveRel s) (ht : IsTransitiveRel t) :
+    IsTransitiveRel (entourageProd s t) :=
+  fun _ _ _ h h' ↦ ⟨hs h.left h'.left, ht h.right h'.right⟩
+
+lemma IsUltraUniformity.comap {u : UniformSpace Y} (h : IsUltraUniformity Y) (f : X → Y) :
+    @IsUltraUniformity _ (u.comap f) := by
+  letI := u.comap f
+  constructor
+  refine (h.hasBasis.comap (Prod.map f f)).to_hasBasis' ?_ ?_
+  · simp only [uniformity_comap, Filter.mem_comap, id_eq, and_imp]
+    intro s hs hss hst
+    exact ⟨_, ⟨⟨s, hs, Set.preimage_mono (subset_refl _)⟩, hss.preimage_map _, hst.preimage_map _⟩,
+      subset_refl _⟩
+  · aesop
```

Comment:
This kind of use case is precisely why I made you add the `mk_of_hasBasis` constructor in the last PR: you can avoid the `to_hasBasis'` thing entirely like this:
```suggestion
lemma IsTransitiveRel.preimage_prodMap {s : Set (Y × Y)} {f : X → Y}
    (hs : IsTransitiveRel s) :
    IsTransitiveRel ((Prod.map f f) ⁻¹' s) :=
  fun x y z ↦ @hs (f x) (f y) (f z)

lemma IsUltraUniformity.comap {u : UniformSpace Y} (h : IsUltraUniformity Y) (f : X → Y) :
    @IsUltraUniformity _ (u.comap f) := by
  letI := u.comap f
  refine .mk_of_hasBasis (h.hasBasis.comap (Prod.map f f)) ?_ ?_
  · exact fun U ⟨_, hU, _⟩ ↦ hU.preimage_map f
  · exact fun U ⟨_, _, hU⟩ ↦ hU.preimage_prodMap
```

(The extra lemma is independent from what I want to highlight)

Reply:
Oh and you had already added it :facepalm:

Reply:
(The same trick should work for `inf` and `iInf`)